### PR TITLE
Towards #2709 - Allow port instead of portIndex for health checks

### PR DESCRIPTION
**Pull request by kolloch**
* Improved health check documentation
  * Replaced JInt with scala Int in HealthCheck case class
  * Automatically add `"portIndex": 0` when necessary to preserve backwards compatibility

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala [hunk 1]
```diff
@@ -446,6 +447,26 @@ trait V2Formats {
...
+    *
+    * In the past, healthCheck.portIndex was required and had a default value 0. When we introduced healthCheck.port, we
+    * made it optional (also with ip-per-container in mind) and we have to re-add it in cases where it makes sense.
+    */
+  private[this] def addHealthCheckPortIndexIfNecessary(v2App: V2AppDefinition): V2AppDefinition = {
+    val hasPortMappings = v2App.container.exists(_.docker.exists(_.portMappings.exists(_.nonEmpty)))
+    val portIndexesMakeSense = v2App.ports.nonEmpty || hasPortMappings
+    if (portIndexesMakeSense) {
+      v2App.copy(healthChecks = v2App.healthChecks.map { healthCheck =>
+        if (healthCheck.port.isEmpty && healthCheck.portIndex.isEmpty && healthCheck.protocol != Protocol.COMMAND) {
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: Can you extract those checks as flag: `val needPortIndex = x&y&z ` and then have only one if: `if (needPortIndex && portIndexesMakeSense) app.copy(...)`

File: src/main/scala/mesosphere/marathon/health/HealthCheck.scala [hunk 2]
```diff
@@ -54,7 +58,13 @@ case class HealthCheck(
       path =
         if (proto.hasPath) Some(proto.getPath) else None,
       protocol = proto.getProtocol,
-      portIndex = proto.getPortIndex,
+      portIndex =
+        if (proto.hasPortIndex)
+          Some(proto.getPortIndex)
+        else if (!proto.hasPort && proto.getProtocol != Protocol.COMMAND)
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/health/HealthCheck.scala]: Can we run into this situation?

File: src/test/scala/mesosphere/marathon/health/HealthCheckTest.scala [hunk 3]
```diff
@@ -31,12 +30,35 @@ class HealthCheckTest extends MarathonSpec {
...
     assert(60 == proto.getIntervalSeconds)
     assert(0 == proto.getMaxConsecutiveFailures)
+    assert(!proto.hasPort)
+  }
+
+  test("ToProtoOverridePort") {
+    val healthCheck = HealthCheck(
+      path = Some("/health"),
+      protocol = Protocol.HTTP,
+      portIndex = Some(0),
```
Comment (aquamatthias) [inline src/test/scala/mesosphere/marathon/health/HealthCheckTest.scala]: Isn't that something we forbid to create: portIndex and port?

**Comment by aquamatthias** [comment]
Only minor comments. Looks good! :+1: 
Please address and :ship:

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala (hunk 1 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: You mean like this?

```scala

  /**
    * Ensure backwards compatibility by adding portIndex to health checks when necessary.
    *
    * In the past, healthCheck.portIndex was required and had a default value 0. When we introduced healthCheck.port, we
    * made it optional (also with ip-per-container in mind) and we have to re-add it in cases where it makes sense.
    */
  private[this] def addHealthCheckPortIndexIfNecessary(v2App: V2AppDefinition): V2AppDefinition = {
    val hasPortMappings = v2App.container.exists(_.docker.exists(_.portMappings.exists(_.nonEmpty)))
    val portIndexesMakeSense = v2App.ports.nonEmpty || hasPortMappings
    v2App.copy(healthChecks = v2App.healthChecks.map { healthCheck =>
        def needsDefaultPortIndex = healthCheck.port.isEmpty && healthCheck.portIndex.isEmpty && healthCheck.protocol != Protocol.COMMAND
        if (portIndexesMakeSense && needsDefaultPortIndex) healthCheck.copy(portIndex = Some(0))
        else healthCheck
    })
  }
```

Slightly less efficient but more readable, I like it.

File: src/main/scala/mesosphere/marathon/health/HealthCheck.scala (hunk 2 above)
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/health/HealthCheck.scala]: We are not sure. We might be for backwards compatibility. As written below, portIndex = 0 was the default int protobufs. We don't know if we can get into this situation by a combination of using an old Marathon version and migrations so we thought that we better cover it.

File: src/test/scala/mesosphere/marathon/health/HealthCheckTest.scala (hunk 3 above)
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/health/HealthCheckTest.scala]: Yes, we do that. This test only covers the serialization. Should we separate this to two different tests?

**Comment by kolloch** [comment]
@aquamatthias I hopefully addressed all your concerns in an additional commit. I can squash our two commits if you are happy (I would like to preserve Evans to give him credit where it's due).

**Comment by aquamatthias** [comment]
Thanks! LGTM please squash and merge.

**Comment by kolloch** [comment]
It is squashed! :)  We will not squash Evans patch.

**Comment by aquamatthias** [comment]
Looks good. Thanks!

**Pull request merged**